edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -23,51 +25,50 @@ public class StoringAsyncEndpointExecutor : IEndpointExecutor
         readonly AsyncEndpointExecutorOptions options;
-        readonly EndpointExecutorFsm machine;
         readonly CancellationTokenSource cts = new CancellationTokenSource();
+        readonly ICheckpointerFactory checkpointerFactory;
+        readonly EndpointExecutorConfig config;
+        AtomicReference<OrderedDictionary> prioritiesToFsms;
+        EndpointExecutorFsm lastUsedFsm;
 
         public StoringAsyncEndpointExecutor(
             Endpoint endpoint,
-            IList<uint> priorities,
-            ICheckpointer checkpointer,
+            ICheckpointerFactory checkpointerFactory,
             EndpointExecutorConfig config,
             AsyncEndpointExecutorOptions options,
             IMessageStore messageStore)
         {
-            Preconditions.CheckNotNull(endpoint);
-            Preconditions.CheckNotNull(config);
-            Preconditions.CheckNotNull(priorities);
-            Preconditions.CheckArgument(priorities.Count != 0);
-            this.checkpointer = Preconditions.CheckNotNull(checkpointer);
+            this.Endpoint = Preconditions.CheckNotNull(endpoint);
+            this.checkpointerFactory = Preconditions.CheckNotNull(checkpointerFactory);
+            this.config = Preconditions.CheckNotNull(config);
             this.options = Preconditions.CheckNotNull(options);
-            this.machine = new EndpointExecutorFsm(endpoint, checkpointer, config);
             this.messageStore = messageStore;
             this.sendMessageTask = Task.Run(this.SendMessagesPump);
-
-            this.UpdatePriorities(priorities);
+            this.prioritiesToFsms = new AtomicReference<OrderedDictionary>(new OrderedDictionary());
         }
 
-        public Endpoint Endpoint => this.machine.Endpoint;
+        public Endpoint Endpoint { get; }
 
-        public EndpointExecutorStatus Status => this.machine.Status;
+        public EndpointExecutorStatus Status => this.lastUsedFsm.Status;
 
         public async Task Invoke(IMessage message, uint priority, uint timeToLiveSecs)
         {
-            // TODO: 6099894 - Update StoringAsyncEndpointExecutor message enqueue logic to be aware of priorities
             try
             {
-                if (this.closed)
-                {
-                    throw new InvalidOperationException($"Endpoint executor for endpoint {this.Endpoint} is closed.");
+                if (this.closed)
+                {
+                    throw new InvalidOperationException($"Endpoint executor for endpoint {this.Endpoint} is closed.");
                 }
 
                 using (MetricsV0.StoreLatency(this.Endpoint.Id))
                 {
-                    IMessage storedMessage = await this.messageStore.Add(this.Endpoint.Id, message);
-                    this.checkpointer.Propose(storedMessage);
+                    // Get the checkpointer corresponding to the queue for this priority
+                    OrderedDictionary snapshot = this.prioritiesToFsms;
```

Comment:
Have you looked at SortedDictionary? It's typed, and it will actually order the dictionary by the sorted key values (priorities)

Reply:
will look into it

Reply:
Yeah, I would love to be able to get rid of the casts if possible.

Reply:
changed to use SortedDictionary, much nicer